### PR TITLE
fix(controller): Do not panic on nil output value. Fixes #3505

### LINKED ISSUE #3505: Panic on addOutputsToLocalScope
Checklist:

 * [x] I've included the version.
 * [ ] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**: 
I'm not sure if stable is also non-viable or is it just me, but I'm getting
```
runtime error: invalid memory address or nil pointer dereference" stack="goroutine 343 [running]:
runtime/debug.Stack(0xc05fda1f51, 0x173d180, 0x28720d0)
  /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).operate.func2(0xc0023d40b0)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:164 +0xc4
panic(0x173d180, 0x28720d0)
  /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).addOutputsToLocalScope(0xc0023d40b0, 0xc002a460a0, 0x15, 0xc0024fa540, 0xc002071a90)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:2072 +0x493
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).buildLocalScope(0xc0023d40b0, 0xc002071a90, 0xc002a460a0, 0x15, 0xc0023ddb80)
 /go/src/github.com/argoproj/argo/workflow/controller/operator.go:2050 +0xaa
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).buildLocalScopeFromTask(0xc0023d40b0, 0xc0018ddea0, 0xc001001ce0, 0x0, 0x0, 0x0)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:475 +0x24c
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).resolveDependencyReferences(0xc0023d40b0, 0xc0018ddea0, 0xc001001ce0, 0xf, 0x0, 0x0)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:485 +0x5d
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc0023d40b0, 0xc0018ddea0, 0xc003465870, 0xf)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:372 +0x7e6
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeDAG(0xc0023d40b0, 0xc002520160, 0x1f, 0xc0024faa00, 0xc0005167b0, 0x25, 0xc0048778c0, 0x1bba5a0, 0xc0023d4160, 0xc00243def0, ...)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:231 +0x2d6
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeTemplate(0xc0023d40b0, 0xc002520160, 0x1f, 0x1bba5a0, 0xc0023d4160, 0xc0024fa980, 0x0, 0x0, 0x0, 0x0, ...)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1517 +0x1de4
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).operate(0xc0023d40b0)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:290 +0x125d
github.com/argoproj/argo/workflow/controller.(*WorkflowController).processNextItem(0xc000097c00, 0x0)
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:433 +0x8c2
github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker(0xc000097c00)
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:359 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc002e4f020)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002e4f020, 0x3b9aca00, 0x0, 0x1, 0xc0000abc80)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc002e4f020, 0x3b9aca00, 0xc0000abc80)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:88 +0x4d
created by github.com/argoproj/argo/workflow/controller.(*WorkflowController).Run
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:172 +0x975
```
with the stable build.

Not sure if there's any change I need to make in my workflow definition.

**What you expected to happen**: no panic

**How to reproduce it (as minimally and precisely as possible)**: 

**Anything else we need to know?**: I upgraded from 2.8.0 to. stable and all workflows that involve output passing failed with this panic.

**Environment**:

- Argo version: 
```
$ argo version
argo: v2.8.0
  BuildDate: 2020-05-11T22:55:43Z
  GitCommit: 8f696174746ed01b9bf1941ad03da62d312df641
  GitTreeState: clean
  GitTag: v2.8.0
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2020-02-13T18:08:14Z"
  compiler: gc
  gitCommit: 06ad960bfd03b39c8310aaf92d1e7c12ce618213
  gitTreeState: clean
  gitVersion: v1.17.3
  goVersion: go1.13.8
  major: "1"
  minor: "17"
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-05-29T15:41:17Z"
  compiler: gc
  gitCommit: b20d34dbc46f336ed52ea4c01356da9952f0b8fc
  gitTreeState: clean
  gitVersion: v1.15.11-gke.17
  goVersion: go1.12.17b4
  major: "1"
  minor: 15+
  platform: linux/amd64
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #3505

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -2081,32 +2081,22 @@ func (woc *wfOperationCtx) buildLocalScope(scope *wfScope, prefix string, node *
...
-		if scope != nil {
-			scope.addParamToScope(key, *outputs.ExitCode)
-		}
+		scope.addParamToScope(fmt.Sprintf("%s.exitCode", prefix), *outputs.ExitCode)
 	}
 	for _, param := range outputs.Parameters {
-		key := fmt.Sprintf("%s.outputs.parameters.%s", prefix, param.Name)
-		if scope != nil {
-			scope.addParamToScope(key, param.Value.String())
+		if param.Value != nil {
```
Comment (alexec) [inline workflow/controller/operator.go:2094]: this is the actual bug fix

File: workflow/controller/operator_test.go [hunk 2]
```diff
@@ -4021,6 +4021,54 @@ status:
...
+
+	podInterface := controller.kubeclientset.CoreV1().Pods("my-ns")
+	list, err := podInterface.List(metav1.ListOptions{})
+	assert.NoError(t, err)
+	// two items -> two pods
+	assert.Len(t, list.Items, 1)
+
+	// mark all pods good
+	for _, pod := range list.Items {
+		pod.Status.Phase = apiv1.PodSucceeded
```
Comment (alexec) [inline workflow/controller/operator_test.go:4061]: @sarabala1979 @simster7 can I highlight this test? I don't think other tests do this, but it is possible to 

1. Create a `woc`
2. `woc.operate()`
3. Set the state of the pods created - in this example I set the phase and outputs.
4. `woc.operate()`
5. assert on workflow state

This means you CAN write a unit test that is as powerful as a functional e2e test - but faster and more robust.

Comment (sarabala1979) [reply workflow/controller/operator_test.go:4159]: Please the existing function for `makePodsPhase`

Comment (alexec) [reply workflow/controller/operator_test.go:4266]: as I though, `makePodsPhase` does not support outputs

I'm starting to think of something like this.

```golang

	cancel, f := newFixture()
	defer cancel()
	f.Create(`
apiVersion: v1alpha1
kind: Workflow
metadata:
  name: my-wf
  namespace: my-ns
spec:
  entrypoint: main
  templates:
  - name: main
    dag:
      tasks:
      - name: step-1
        template: child
  - name: child
    container:
      image: my-image
    outputs:
      parameters:
      - name: my-param
        valueFrom: 
          path: /my-path
`).
		Reconcile().
		Merge(`
kind: Pod
metadata:
  name: ...
  annotations:
    workflows.argoproj.io/outputs: "{...}"
`).
		Reconcile().
		Expect(func(wf *wfv1.Workflow) {
			assert.Equal(t, wfv1.NodeSucceeded, wf.Status.Phase)
		
		})

```

**Pull request merged**